### DISCUSSION TITLE
How to run only detection model to get the bounding boxes. (Handwritten text)

### DISCUSSION THREAD

**Discussion by sanjay-nit** [mindee/doctr#1734] [Q&A] [answered]
Is there any way to run only the detection model and get the bounding boxes.
So, Idea is to run it for hand written documents and it seems like TrOCR works well on handwritten but it only works of one liner text images. So, I'm thinking if I can run only the detection model to get the line/word bounding boxes so that I can crop it and send it to TrOCR. Thanks!

**Comment by felixdittrich92** [comment, marked as answer]
Hi @sanjay-nit :wave:,
Sure the `ocr_predictor` instance is at the end only a wrapper around the `detection_predictor` / `recognition_predictor` and `crop_orientation_predictor /  page_orientation_predictor`.

Here you go (example code):

```
import requests
import cv2
import numpy as np

from doctr.io import DocumentFile
from doctr.models import detection_predictor
from doctr.utils.geometry import detach_scores


# Convert relative coordinates to absolute pixel values
def _to_absolute(geom, img_shape: tuple[int, int]) -> list[list[int]]:
    h, w = img_shape
    if len(geom) == 2:  # Assume straight pages = True -> [[xmin, ymin], [xmax, ymax]]
        (xmin, ymin), (xmax, ymax) = geom
        xmin, xmax = int(round(w * xmin)), int(round(w * xmax))
        ymin, ymax = int(round(h * ymin)), int(round(h * ymax))
        return [[xmin, ymin], [xmax, ymin], [xmax, ymax], [xmin, ymax]]
    else:  # For polygons, convert each point to absolute coordinates
        return [[int(point[0] * w), int(point[1] * h)] for point in geom]


url = "https://www.francetvinfo.fr/pictures/uGwaNE-aJq7zHLhZJdzdCd9nyjE/1200x900/2021/03/16/phpCDwGn0.jpg"

det_predictor = detection_predictor(
    arch="fast_base",
    pretrained=True,
    assume_straight_pages=True,
    symmetric_pad=True,
    preserve_aspect_ratio=True,
) #.cuda().half()  # Uncomment this line if you have a GPU

det_predictor.model.postprocessor.bin_thresh = 0.3
det_predictor.model.postprocessor.box_thresh = 0.65

docs = DocumentFile.from_images([requests.get(url).content])
results = det_predictor(docs)

image = cv2.imdecode(np.frombuffer(requests.get(url).content, np.uint8), cv2.IMREAD_COLOR)

for doc, res in zip(docs, results):
    img_shape = (doc.shape[0], doc.shape[1])
    # Detach the probability scores from the results
    detached_coords, prob_scores = detach_scores([res.get("words")])

    for i, coords in enumerate(detached_coords[0]):
        coords = coords.reshape(2, 2).tolist() if coords.shape == (4, ) else coords.tolist()

        # Convert relative to absolute pixel coordinates
        points = np.array(_to_absolute(coords, img_shape), dtype=np.int32).reshape((-1, 1, 2))

        # Draw the bounding box on the image
        cv2.polylines(image, [points], isClosed=True, color=(255, 0, 0), thickness=2)

    # Save the modified image with bounding boxes
    cv2.imwrite("output.jpg", image)
 ```

**Reply by sanjay-nit** [reply]
Thanks @felixdittrich92 for your quick reply. You are really a life saver, man. 
Well, I don't see `detach_scores` method in `doctr.utils.geometry.py` file. FYI: using `python-doctr==0.8.1`
![image](https://github.com/user-attachments/assets/66d863f7-1c52-4a62-9462-6b815d70d9b4)

**Reply by felixdittrich92** [reply]
@sanjay-nit I suggest to upgrade to `0.9.0` because there was a fix included which solved a significant bug by using the standalone `detection_predictor` :)

ref.: https://github.com/mindee/doctr/pull/1627

**Comment by sanjay-nit** [comment]
Hi @felixdittrich92,

This may not be directly related to this discussion, but I wanted to express my satisfaction with the auto-rotating pages feature.

I have some rotated images on which I want to apply OCR, and I also need to straighten the pages. I'm pleased with the results when I use `assume_straight_pages=False` and `straighten_pages=True` together in the `ocr_predictor`. However, when I export the result, I don't receive the page rotation angle, even though it works beautifully when I use the `result.show()` method.

Is it possible to get the rotation angle? Additionally, it would be great if I could obtain the auto-rotated image that the OCR model processed so I can apply the bounding boxes generated by the `result.export()` method.

Thank you!

**Comment by felixdittrich92** [comment]
Hi @sanjay-nit :wave:,

Glad to hear you like it :hugs: 

You can pass `detect_orientation=True,` to the `ocr_predictor` to get the estimated angle :)

To get the straightened page you can do the following:

```
result = model(doc)

# get straightened pages
corrected_pages = [page.page for page in result.pages]
```

This results in a list of numpy arrays which can easily saved with `opencv` :)

Hope this helps :+1: 

Best regards,
Felix

**Comment by felixdittrich92** [comment]
@sanjay-nit Do you deal with "random" rotated documents and images where the text can also have any direction (horizontal, vertical) or it's more like small rotations (range: -45 to 45 degrees) and only horizontal text ? :)

**Reply by sanjay-nit** [reply]
@felixdittrich92 Rotations can be randoms. For eg user is taking photo of documents using smartphones and sometimes pages might get rotated by -90, +90 or 180 or it can be any degree. We are trying to see if there is a way to automatically detect orientations.

**Reply by felixdittrich92** [reply]
:+1: 
(Currently only available on main branch)
We added also support to train custom orientation models if you should see poor results you can train / fine tune `page` and/or `crop` orientation models to your own needs.
Ref.: https://mindee.github.io/doctr/latest/using_doctr/custom_models_training.html#custom-orientation-classification-models

**Reply by sanjay-nit** [reply]
Thanks @felixdittrich92,

1. Is there a way to get a selectable/searchable PDF when we load a scanned PDF and process it through DocTR OCR engine?
2. Also, can we get the character-level bounding boxes from the OCR output?

The idea is that if we are able to get a selectable PDF back from the OCR output, then we are good to go. But in case it's not possible, I'm thinking of writing an [OCRmyPDF](https://github.com/ocrmypdf/OCRmyPDF) plugin which uses DocTR for better OCR results rather than its default Tesseract. However, for this to work, DocTR should return similar outputs as Tesseract. Thanks again!! :-)

**Reply by felixdittrich92** [reply]
Hi @sanjay-nit :wave:,

1. It's possible but an OCRmyPDF plugin would be a nice addition and i think provide some improvements over our solution:
https://www.mindee.com/blog/create-ocrized-pdfs-in-2-steps
https://colab.research.google.com/github/mindee/notebooks/blob/main/doctr/export_as_pdfa.ipynb
2. That's not possible atm :)

You can get the results directly in hocr format (does not work with rotations atm - only straight boxes supported)

```
xml_output = result.export_as_xml()
for output in xml_output:
    xml_bytes_string = output[0]
    xml_element = output[1]
```

```
<?xml version="1.0" encoding="UTF-8"?>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
  <head>
    <title>docTR - hOCR</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="ocr-system" content="doctr 0.5.0" />
    <meta name="ocr-capabilities" content="ocr_page ocr_carea ocr_par ocr_line ocrx_word" />
  </head>
  <body>
    <div class="ocr_page" id="page_1" title="image; bbox 0 0 3456 3456; ppageno 0" />
    <div class="ocr_carea" id="block_1_1" title="bbox 857 529 2504 2710">
      <p class="ocr_par" id="par_1_1" title="bbox 857 529 2504 2710">
        <span class="ocr_line" id="line_1_1" title="bbox 857 529 2504 2710; baseline 0 0; x_size 0; x_descenders 0; x_ascenders 0">
          <span class="ocrx_word" id="word_1_1" title="bbox 1552 540 1778 580; x_wconf 99">Hello</span>
          <span class="ocrx_word" id="word_1_2" title="bbox 1782 529 1900 583; x_wconf 99">XML</span>
          <span class="ocrx_word" id="word_1_3" title="bbox 1420 597 1684 641; x_wconf 81">World</span>
        </span>
      </p>
    </div>
  </body>
</html>
```

**Reply by sanjay-nit** [reply]
Thanks a lot @felixdittrich92 💟 
I just tested it and with few/minimal changes it is working fine for images. And also for single page PDF.
In case of multipage PDF, do I need to process each page one by one(since it requires image) and create separate PDF for each page; and finally use some PDF merger tool? Is this the way? Any suggestions @felixdittrich92 ?

**Reply by felixT2K** [reply]
@sanjay-nit Which one have you tested ? The blog post ? 
With some small changes it should also work for pdf files or multible images to create one pdf/a from all without the need of any merging :)
The OCRmyPDF plugin would still be cool 😅 Do you know if OCRmyPDF can also handle creating pdf/a files from images/pdfs which contains rotated text ?

**Reply by felixT2K** [reply]
I think we should update our tutorial notebook also to use OCRmyPDF instead of the own implementation 🤔 Is tesseract a fixed dependency in OCRmyPDF or optional ?

**Reply by sanjay-nit** [reply]
> @sanjay-nit Which one have you tested ? The blog post ? With some small changes it should also work for pdf files or multible images to create one pdf/a from all without the need of any merging :) The OCRmyPDF plugin would still be cool 😅 Do you know if OCRmyPDF can also handle creating pdf/a files from images/pdfs which contains rotated text ?

Hi @felixT2K 
Yes I tested the blogpost code with image and single page pdf. 
Yes, I tested OCRmyPDF for auto page rotation and what I found that most of the times it doesn't work perfectly.

**Reply by sanjay-nit** [reply]
> I think we should update our tutorial notebook also to use OCRmyPDF instead of the own implementation 🤔 Is tesseract a fixed dependency in OCRmyPDF or optional ?
@felixT2K 
That would be great 😄 😄 
They use tesseract internally but have option to write plugins. [OCRmyPDF plugin](https://ocrmypdf.readthedocs.io/en/latest/plugins.html#ocr-engine)

**Reply by felixdittrich92** [reply]
@sanjay-nit https://github.com/mindee/notebooks/pull/20
CC @odulcy-mindee

**Comment by hanshupe007** [comment]
Is there any TrOCR integration planned for docTr or other support for handwritten texts?

**Reply by felixdittrich92** [reply]
Hi @hanshupe007 👋🏼,

TrOCR support isn't planned because we have architectures like `parseq` | `master` | `sar_resnet31` which definitely have the capacity to handle handwritten recognition :)
But you can easily combine docTR's `detection_predictor` with TrOCR on your own.
In general handwritten support is planned but it's a long term issue because we don't have such a dataset unfortunately.

**Reply by hanshupe007** [reply]
Thanks! Before trying it out, does it mean that the pretrained models mentioned above work on handwritten texts, or only if I would train in on a related dataset?

**Reply by felixdittrich92** [reply]
For some good readable (printed writing) it should already work partially with the current pretrained models.
For a "robust" handwritten model which can also handle cursive writing you need a dataset to fine tune on.

**Reply by felixdittrich92** [reply]
I found an dataset which looks interesting for such first experiments:

https://huggingface.co/datasets/shreyansh1347/GNHK-Synthetic-OCR-Dataset/viewer/default/test

This would only require to convert into the docTR required format :)

**Reply by hanshupe007** [reply]
Fyi, did some first tests with TrOCR on handwritten texts and it looks like it's performing very poorly, also compared to parseq. Thinking now about the following approaches:
.) fine-tuning parseq on handwritten texts
.) exploring open sourced multimodal-models, maybe in combination with a detection model

if anyone has already experience with handwritten texts, it would be appreciated : )

**Reply by kkaarrss** [reply]
> Fyi, did some first tests with TrOCR on handwritten texts and it looks like it's performing very poorly, also compared to parseq. Thinking now about the following approaches: .) fine-tuning parseq on handwritten texts .) exploring open sourced multimodal-models, maybe in combination with a detection model
> 
> if anyone has already experience with handwritten texts, it would be appreciated : )

Qwen2 VL (only 7B) is the best open source for handwriting. At half precision it fits on a single 24GB GPU. I am using parseq + doctr to orient the incoming documents because Qwen2 does not perform very well when the document is on it's side or upside down. I also do some standard OCR preprocessing of the image, although it does not matter too much.

You can test this setup here: https://easymarks.ai/handwriting-to-text

**Reply by hanshupe007** [reply]
Cool, have you experienced any issues that it completely misses texts (like columns in a table). Or is there any specific prompt to use? Would it make sense to crop words first using a detection model?

**Reply by sanjay-nit** [reply]
> Fyi, did some first tests with TrOCR on handwritten texts and it looks like it's performing very poorly, also compared to parseq. Thinking now about the following approaches: .) fine-tuning parseq on handwritten texts .) exploring open sourced multimodal-models, maybe in combination with a detection model
> 
> if anyone has already experience with handwritten texts, it would be appreciated : )

I hope you are passing the single line text images and the correct model checkpoint.

**Reply by kkaarrss** [reply]
Yes, once on a paper without lines and the written text was not straight, it skipped a line. I was not able to fix that with prompt engineering, but 7B maybe is too stupid for that. 72B might respond better to prompt engineering although the actual OCR capabilities are exactly the same between 72B and 7B. But if the input is not too crazy it usually gets it pretty well.

**Reply by sanjay-nit** [reply]
> Yes, once on a paper without lines and the written text was not straight, it skipped a line. I was not able to fix that with prompt engineering, but 7B maybe is too stupid for that. 72B might respond better to prompt engineering although the actual OCR capabilities are exactly the same between 72B and 7B. But if the input is not too crazy it usually gets it pretty well.

I hope you are not referring the trocr model here, since it's not that big model 😄 ? I'm talking about the trocr correct checkpoint from microsoft.

**Reply by kkaarrss** [reply]
No, Qwen2 VL, but will check out trocr to compare.
